src/stan/services/optimize/bfgs.hpp

```diff
@@ -0,0 +1,202 @@
+       * @param[out] interrupt interrupt callback to be called every iteration
+       * @param[out] message_writer output for messages
+       * @param[out] init_writer Writer callback for unconstrained inits
+       * @param[out] parameter_writer output for parameter values
+       * @return stan::services::error_codes::OK if successful
+       */
+      template <class Model>
+      int bfgs(Model& model,
+               stan::io::var_context& init,
+               unsigned int random_seed,
+               unsigned int chain,
+               double init_radius,
+               double init_alpha,
+               double tol_obj,
+               double tol_rel_obj,
+               double tol_grad,
+               double tol_rel_grad,
+               double tol_param,
+               int num_iterations,
+               bool save_iterations,
+               int refresh,
+               callbacks::interrupt& interrupt,
+               callbacks::writer& message_writer,
+               callbacks::writer& init_writer,
+               callbacks::writer& parameter_writer) {
+        boost::ecuyer1988 rng = stan::services::util::rng(random_seed, chain);
+
+        std::vector<int> disc_vector;
+        std::vector<double> cont_vector
+          = stan::services::util::initialize(model, init, rng, init_radius,
+                                             false,
+                                             message_writer, init_writer);
+
+        std::stringstream bfgs_ss;
+        typedef stan::optimization::BFGSLineSearch
+          <Model, stan::optimization::BFGSUpdate_HInv<> > Optimizer;
+        Optimizer bfgs(model, cont_vector, disc_vector, &bfgs_ss);
+        bfgs._ls_opts.alpha0 = init_alpha;
+        bfgs._conv_opts.tolAbsF = tol_obj;
+        bfgs._conv_opts.tolRelF = tol_rel_obj;
+        bfgs._conv_opts.tolAbsGrad = tol_grad;
+        bfgs._conv_opts.tolRelGrad = tol_rel_grad;
+        bfgs._conv_opts.tolAbsX = tol_param;
+        bfgs._conv_opts.maxIts = num_iterations;
+
+        double lp = bfgs.logp();
+
+        std::stringstream msg;
+        msg << "Initial log joint probability = " << lp;
+        message_writer(msg.str());
+
+        std::vector<std::string> names;
+        names.push_back("lp__");
+        model.constrained_param_names(names, true, true);
+        parameter_writer(names);
+
+        if (save_iterations) {
+          std::vector<double> values;
+          msg.str("");
+          model.write_array(rng, cont_vector, disc_vector, values,
+                            true, true, &msg);
+          if (msg.str().length() > 0)
+            message_writer(msg.str());
+
+          values.insert(values.begin(), lp);
+          parameter_writer(values);
+        }
+        int ret = 0;
+
+        while (ret == 0) {
+          interrupt();
+          if (refresh > 0
+              && (bfgs.iter_num() == 0
+                  || (bfgs.iter_num() + 1 % refresh == 0)))
+            message_writer("    Iter "
```

Comment:
No. Of course not. But I'm trying hard not to fix everything that's wrong with our whole code base or we'll never get this in. This has been in since it was written -- it bothers me too, but I'm really trying to fix the things that need to be fixed. We can walk through the code base later.